examples/exchange-sudt-for-ckb/src/lib.ts

```diff
@@ -0,0 +1,327 @@
+      capacity: CKBChangeCapacity.toHexString(),
+      lock: holderAccountInfo.lockScript,
+    },
+    data: "0x",
+  };
+  txSkeleton = txSkeleton.update("outputs", (outputs) => outputs.push(CKBChangeOutput));
+
+  txSkeleton = txSkeleton.update("witnesses", (witnesses) => {
+    const placeholderWitness = {
+      lock:
+        "0x0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
+    };
+
+    const witnessArgs = bytes.hexify(blockchain.WitnessArgs.pack(placeholderWitness));
+
+    for (let index = 0; index < txSkeleton.inputs.size; index++) {
+      witnesses = witnesses.push(witnessArgs);
+    }
+
+    return witnesses;
+  });
+
+  txSkeleton = await payFeeByFeeRate(txSkeleton, [holderAccountInfo.address], 1000, undefined, { config: AGGRON4 });
+
+  // STEP2: sign the transaction skeleton
+  const tx = signTx(txSkeleton, [issuerPrivateKey, holderPrivateKey, issuerPrivateKey]);
+
+  // STEP3: send it to testnet
+  const txHash = await rpc.sendTransaction(tx, "passthrough");
+  return txHash;
+}
+
+/**
+ * sign a transaction
+ */
+export function signTx(txSkeleton: helpers.TransactionSkeletonType, privateKey: string[]) {
+  txSkeleton = commons.common.prepareSigningEntries(txSkeleton);
+  const signatures = txSkeleton.get("signingEntries").map(({ message }, index) => {
+    return hd.key.signRecoverable(message, privateKey[index]);
+  });
+  const tx = helpers.sealTransaction(txSkeleton, signatures.toJSON());
+  return tx;
+}
+
+export function calculateSUDTAmountSum(cells: Cell[]) {
+  let amount = BI.from(0);
+  for (const cell of cells) {
+    if (cell.cellOutput.type?.codeHash === AGGRON4.SCRIPTS.SUDT.CODE_HASH) {
+      amount = amount.add(number.Uint128LE.unpack(cell.data));
+    }
+  }
+
+  return amount;
+}
+
+export function calculateCKBSum(cells: Cell[]) {
+  let amount = BI.from(0);
+
+  for (const cell of cells) {
+    amount = amount.add(cell.cellOutput.capacity);
+  }
+
+  return amount;
+}
+
+/**
+ * collect all SUDT cells and calculate their data sum
+ * @param address SUDT Issuer address
+ * @returns the amount of SUDT in the address
+ */
+export async function fetchSUDTBalance(address: string) {
+  const collector = indexer.collector({
+    lock: helpers.parseAddress(address, { config: AGGRON4 }),
+    type: {
+      script: {
+        codeHash: AGGRON4.SCRIPTS.SUDT.CODE_HASH,
+        hashType: AGGRON4.SCRIPTS.SUDT.HASH_TYPE,
+        args: "0x",
+      },
+    },
+  });
+  let amount = BI.from(0);
+
+  for await (const cell of collector.collect()) {
+    amount = amount.add(number.Uint128LE.unpack(cell.data));
+  }
+  return amount;
+}
+
+/**
+ * collect all cells and calculate their capacity sum
+ * @param address
+ * @returns the capacity owns by the address
+ */
+export async function fetchCKBBalance(address: string) {
+  const collector = indexer.collector({ lock: helpers.parseAddress(address, { config: AGGRON4 }) });
```

Comment:
Try using collector like: `const collector = indexer.collector({ lock: helpers.parseAddress(address, { config: AGGRON4 }), outputDataLength: ["0x0", "0x1"] });` to avoid collecting sUDT cells when calculating CKB balance

Reply:
I find a better way to exclude SUDT cells.
```
indexer.collector({ lock: helpers.parseAddress(address, { config: AGGRON4 }), type: "empty" })
``